### ISSUE TITLE
Implementation of automatic entropy temperature tuning(alpha loss)

### ISSUE THREAD

**Issue by Maggern3** [rail-berkeley/softlearning#163]
https://github.com/rail-berkeley/softlearning/blob/46f14436f62465a02b99f431bbcf57a7fa0fd09d/softlearning/algorithms/sac.py#L254-L255
The implementation of the alpha loss seems to vary from the formula definition, formula 18 in the paper Haarnoja et al., 2019. Is this a bug?

**Comment by hartikainen** [comment]
Here's an older discussion about the same topic: https://github.com/rail-berkeley/softlearning/issues/149. Hopefully that answers the question. Please let me know if anything is left unclear!

**Comment by Maggern3** [comment]
Thank you! @hartikainen 

I'm still having trouble. No matter how I implement the temperature it crashes or it goes to 0.0 in a few dozen episodes.

I'm sure there's something wrong with my implementation but I'm clueless as to what. If you could take a look, and you see something I've misunderstood, I'd be happy if you'll let me know.
```
self.expected_entropy = -torch.prod(torch.tensor(action_space.shape).to(self.device)).item() # unsure if this is right for multidiscrete env
print('target entropy', self.expected_entropy) # gives target entropy -4
self.log_alpha = torch.tensor(0.0, requires_grad=True, device=self.device)
self.alpha = self.log_alpha.exp() #0.2#, requires_grad=True, device=self.device)#0.2
self.alpha_optimizer = optim.Adam([self.log_alpha], lr=0.003)
```

```
# my impl based on formula 18 from paper, crashes
#alpha_loss = (-self.alpha * (log_prob - self.expected_entropy).detach()).mean()
# rail-berkeley/softlearning, crashes
#alpha_loss2 = -1.0 * (self.alpha * (log_prob + self.expected_entropy).detach()).mean()
# cyoon1729/Policy-Gradient-Methods, alpha is less than 0.0 in 80 episodes
#alpha_loss3 = (self.log_alpha * (-log_prob - self.expected_entropy).detach()).mean()    
# vitchyr/rlkit, alpha is less than 0.0 in 52 episodes
# p-christ/Deep-Reinforcement-Learning-Algorithms-with-PyTorch alpha is less than 0.0 in 50 episodes
alpha_loss4 = -(self.log_alpha * (next_log_prob_selected_actions + self.expected_entropy).detach()).mean()
self.alpha = self.log_alpha.exp()
self.alpha_optimizer.zero_grad()
alpha_loss4.backward()
self.alpha_optimizer.step() 
```

What would be a good value for expected_entropy or the target entropy for a MultiDiscrete([3 3 2 3]) action space(Obstacle tower environment)?
Depending on how I calculate I get -4, -11 or -54 but I'm a bit clueless to what would be a good value. Ascertaining from the other post you linked, -4 or -11 should work. But right now they're not working. But that could also be another part of my code. 

Also if you could explain the difference in the math between my implementation and yours above I'd appreciate it. Why -1 as a multiplier? To get the gradients moving the right direction? And why log_prob + self.expected_entropy, in the paper it's minus?

Edit, if you need to see full source code it's [here](https://github.com/Maggern3/SAC/blob/Sac19MultiDiscrete/sac.py)

Also another question about the intuition, if it should be different for each state(the entropy), shouldn't we create it's own neural net? How can one tensor encapsulate different entropy values for each different state? Is it achieved through how we use alpha together with the other losses? I still don't get it..

Thanks!

**Comment by hartikainen** [comment]
Hey @Maggern3. Apologies for such a delayed reply.

The heuristic we use in our paper [[1]](https://arxiv.org/pdf/1812.05905.pdf) assumes continuous action spaces. I have never used SAC for discrete action spaces, but the heuristic we use in the paper would not work in such setting because the discrete entropy should be non-negative. What happens if you try using positive values for `self.expected_entropy`?

Regarding the difference between your and e.g. `softlearning` or `rlkit` implementation: as you already noted, it seems like your implementation has a wrong sign for `self.expected_entropy`. If you write open the equations, the ones we use in `softlearning`/`rlkit` should be equivalent with one from the paper [[1]](https://arxiv.org/pdf/1812.05905.pdf). My code for some reason explicitly multiplies by `-1.0` instead of simply negating the value, but in practice these result in same gradients.

If you have multi-dimensional discrete variable, I think the heuristic should consider the size of their product, the same way as we do for continuous action spaces. What the exact heuristic for discrete values would be, I'm not exactly sure. It should be somewhat easy to come up with one though, because for discrete distributions the minimum entropy is 0 and you should be able to compute the maximum entropy in closed form. Then the heuristic could be something like `multiplier * maximum_entropy`. It should probably tested how well that works for different environments. See for example Figure 4 in [[2]](https://arxiv.org/pdf/1812.11103.pdf) for how the entropy heuristic was tested for continuous mujoco environments.

One more high-level note about using SAC in discrete action spaces: given that your action space size is relatively small, it might be worth removing the actor network altogether, and just deriving an implicit policy directly from Q (similarly as you would do in e.g. DQN) by choosing actions based on softmax of the Q values. The reason we need an explicit actor network in continuous case is that it's unfeasible to compute the (soft-) max of the Q values, but in discrete case you don't have such issue.

Hope this clarifies things a bit. Let us know if you have any further questions!

[[1] Haarnoja, T., Zhou, A., Hartikainen, K., Tucker, G., Ha, S., Tan, J., Kumar, V., Zhu, H., Gupta, A., Abbeel, P. and Levine, S., 2018. Soft actor-critic algorithms and applications. arXiv preprint arXiv:1812.05905.](https://arxiv.org/pdf/1812.05905.pdf)
[[2] Haarnoja, T., Ha, S., Zhou, A., Tan, J., Tucker, G. and Levine, S., 2018. Learning to walk via deep reinforcement learning. arXiv preprint arXiv:1812.11103.](https://arxiv.org/pdf/1812.11103.pdf)

**Comment by Maggern3** [comment]
Thank you for some valuable input! I'm trying out some of the changes you suggested now and I will comment back on how it goes.

> One more high-level note about using SAC in discrete action spaces: given that your action space size is relatively small, it might be worth removing the actor network altogether, and just deriving an implicit policy directly from Q (similarly as you would do in e.g. DQN) by choosing actions based on softmax of the Q values. The reason we need an explicit actor network in continuous case is that it's unfeasible to compute the (soft-) max of the Q values, but in discrete case you don't have such issue.

That's an interesting experiment once I get things working. I'm not sure if you saw my code but the code is currently in my opinion overly complex in order to go from continuous actions to multi-discrete actions. I believe there is another factor in play here though, the balancing of variance and bias between policy based methods and value based methods? That is empirically shown to perform better by having separate actor and critic networks. Thoughts on this?

Another question about the intuition, if the entropy should be different for each state, to explore low reward states less and high or unknown reward states more, shouldn't it have it's own neural net? How can one tensor encapsulate different entropy values for all the different states? I still don't understand it. Thanks!

**Comment by Maggern3** [comment]
So that didn't work, alpha goes to 0. You mentioned earlier our math should give the same gradients if we open it up. I think something's going wrong here. Why are we adding log_prob and self.expected_entropy, in the paper [[1]](https://arxiv.org/abs/1812.05905) they are subtracted? 

episode 44 frames 600 rewards 0.00 mean score 0.68 **alpha 2.802596928649634e-45 target entropy 4**

**Comment by Maggern3** [comment]
@hartikainen 

The result in the last post was from using alpha_loss4.

Using my loss function(alpha_loss) I noticed another key difference. I'm using alpha directly where you're using log_alpha. Why is that? In the paper's [[1]](https://arxiv.org/abs/1812.05905) formula 18 it doesn't say log alpha, just alpha and then the log prob.

`
#my impl based on formula 18 from paper, crashes
alpha_loss = (-self.alpha * (next_log_prob_selected_actions - self.expected_entropy).detach()).mean()
#my second implementation based on modifications suggested by authors on github issue 163
alpha_loss = (-self.log_alpha * (next_log_prob_selected_actions + self.expected_entropy).detach()).mean()`

I tried that out above, but since log_alpha is an exponential it's ever decaying and ends up being 0.00 around 50 episodes. That's not what we want, that means zero exploration.
